src/invidious/routes/api/v1/authenticated.cr

```diff
@@ -54,6 +54,60 @@ module Invidious::Routes::API::V1::Authenticated
     env.response.status_code = 204
   end
 
+  def self.get_history(env)
+    env.response.content_type = "application/json"
+    user = env.get("user").as(User)
+
+    page = env.params.query["page"]?.try &.to_i?
+    page ||= 1
+
+    max_results = env.params.query["max_results"]?.try &.to_i?.try &.clamp(0, MAX_ITEMS_PER_PAGE)
+    max_results ||= user.preferences.max_results
+    max_results ||= CONFIG.default_user_preferences.max_results
+
+    if user.watched[(page - 1) * max_results]?
+      watched = user.watched.reverse[(page - 1) * max_results, max_results]
+    end
+    watched ||= [] of String
+
+    return watched.to_json
+  end
+
+  def self.mark_watched(env)
+    user = env.get("user").as(User)
+
+    if !user.preferences.watch_history
+      return error_json(409, "Watch history is disabled in preferences.")
+    end
+
+    id = env.params.url["id"]?.try &.as(String)
+    if !id
+      return error_json(400, "Invalid video id.")
+    end
+
+    Invidious::Database::Users.mark_watched(user, id)
+    env.response.status_code = 204
+  end
+
+  def self.mark_unwatched(env)
+    user = env.get("user").as(User)
+
```

Comment:
In order to save a DB call, you can check if watch history is disabled here

Reply:
This one I left intentionally. If someone would want to remove a video from history (for some reason) they might be stuck if the history got disabled after the video was added. This is mostly a privacy concern, people should be able to remove from history regardless of the setting. But adding a watched video would be guarded by the setting

Reply:
The "clear history" can still be called, so I guess it's fine?